### PR TITLE
Stream.append() fixed #824

### LINKED ISSUE #824: Stream.append produces StackOverFlow
@netzwerg discovered a StackOverflowError: https://github.com/netzwerg/paleo/commit/e962a43fa435aa54211ad17b8c093f5cc0ec0991

### PR DESCRIPTION
**Pull request by ruslansennov**

### CODE REVIEW

File: javaslang/src/main/java/javaslang/collection/Stream.java [hunk 1]
```diff
@@ -1414,28 +1408,19 @@ private Object readResolve() {
...
-     * {@link java.lang.AutoCloseable}.
+     * Non-empty {@code Stream}, consisting of a {@code head}, and {@code tail}.
      *
      * @param <T> Component type of the Stream.
      * @since 1.1.0
      */
-    // DEV NOTE: class declared final because of serialization proxy pattern.
-    // (see Effective Java, 2nd ed., p. 315)
-    final class Cons<T> implements Stream<T>, Serializable {
+    abstract class Cons<T> implements Stream<T> {
```
Comment (ruslansennov) [inline javaslang/src/main/java/javaslang/collection/Stream.java]: @danieldietrich I know you don't like abstract classes in such context, but but here it simplifies the code

File: javaslang/src/main/java/javaslang/collection/Stream.java [hunk 2]
```diff
@@ -1506,113 +1486,160 @@ public String toString() {
             }
             return builder.append(")").toString();
         }
+    }
+}
+
+interface StreamModule {
+
+    final class ConsImpl<T> extends Cons<T> implements Serializable {
```
Comment (ruslansennov) [inline javaslang/src/main/java/javaslang/collection/Stream.java]: we can rename it to something else

**Comment by danieldietrich** [comment]
Just saw your PR, thank you! Need to sleep - will review the changes tomorrow!

File: javaslang/src/test/java/javaslang/collection/StreamTest.java [hunk 3]
```diff
@@ -233,6 +233,14 @@ public void shouldNarrowStream() {
         assertThat(actual).isEqualTo(3);
     }
 
+    // -- append
+
+    @Test
+    public void shouldAppendMillionTimes() {
+        final int bigNum = 1_000_000;
+        assertThat(Stream.range(0, bigNum).foldLeft(Stream.empty(), Stream::append).length()).isEqualTo(bigNum);
```
Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/StreamTest.java]: I'm really impressed by the AppendElements addition! Very cool! :-)

File: javaslang/src/main/java/javaslang/collection/Stream.java (hunk 2 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Stream.java]: The name is ok!

**Comment by danieldietrich** [comment]
Well done! This is one master-piece beside AppendSelf and HAMT :-)

**Pull request merged**